iota-execution/latest/iota-verifier/src/account_auth_verifier.rs

```diff
@@ -0,0 +1,261 @@
+/// before calling the function itself.
+///
+/// A general struct, with no unresolved template arguments:
+/// ```move
+/// public struct Simple has store {
+///   a: u8,
+///   some_vec: vector<ascii::String>
+/// }
+/// ```
+/// is not considered a `pure input` either as it is not a built-in type so it
+/// can't be constructed before the (single) PTB move call itself. On
+/// the contrary std::ascii::String and std::string::String are okay.
+/// On a similar notion a simple `vector<T>` and an `Option<T>` are both also
+/// acceptable as they are built-in move types with rust side counterpart as
+/// long as `T` is recursively `pure` as well.
+fn verify_pure_input_type(
+    module: &CompiledModule,
+    function_type_args: &[AbilitySet],
+    param: &SignatureToken,
+) -> Result<(), String> {
+    use SignatureToken::*;
+
+    match param {
+        U8 | U16 | U32 | U64 | U128 | U256 | Bool | Address => Ok(()),
+        Vector(inner) => verify_pure_input_type(module, function_type_args, inner),
+        Datatype(handle_index) => {
+            let resolved_struct = resolve_struct(module, *handle_index);
+            if resolved_struct == RESOLVED_ASCII_STR
+                || resolved_struct == RESOLVED_UTF8_STR
+                || resolved_struct == RESOLVED_IOTA_ID
+            {
+                Ok(())
+            } else {
+                Err(format!(
+                    "Invalid pure type. A datatype must be a string or an ID, offending argument: {:?}",
+                    param
+                ))
+            }
+        }
+        DatatypeInstantiation(datatype_instance) => {
+            let (idx, type_args) = &**datatype_instance;
+            let resolved_struct = resolve_struct(module, *idx);
+            if resolved_struct == RESOLVED_STD_OPTION && type_args.len() == 1 {
+                verify_pure_input_type(module, function_type_args, &type_args[0])
+            } else {
+                Err(format!(
+                    "Invalid pure type. A datatype instantiation must be an option of pure types, offending argument: {:?}",
+                    param
+                ))
+            }
+        }
+        TypeParameter(idx) => {
+            if function_type_args[*idx as usize].has_key() {
+                Err(format!(
+                    "Invalid pure type. A type parameter cannot have the 'key' ability, offending argument: {:?}",
+                    param
+                ))
+            } else {
+                Ok(())
+            }
+        }
+        Signer => Err(format!(
+            "Invalid pure type. Signer is not a pure type, offending argument: {:?}",
+            param
+        )),
+        Reference(_) => Err(format!(
+            "Invalid pure type. Reference is not a pure type, offending argument: {:?}",
+            param
+        )),
+        MutableReference(_) => Err(format!(
+            "Invalid pure type. MutableReference is not a pure type, offending argument: {:?}",
+            param
```

Comment:
Can do that in all the places in this file
```suggestion
            "Invalid pure type. MutableReference is not a pure type, offending argument: {param:?}"            
```

Reply:
I can, but I don't see why this pattern is any good in general.
Doesn't work in the general case, where you don't have a clean variable name, but an expression `something.get_value()`.

Reply:
Yes doesn't work in all cases, but here are only clean variable names
There is also a (pedantic) clippy lint that suggests that
https://rust-lang.github.io/rust-clippy/master/index.html#uninlined_format_args

Reply:
Ehh, there is some merit to it. Will add it, but it would be better if we just enabled the clippy checks and let that handle this outside of PRs.

Reply:
Fine by me, was just a suggestion :smile: